drink/src/runtime/pallet_contracts_debugging/runtime.rs

```diff
@@ -0,0 +1,79 @@
+use parity_scale_codec::Encode;
+use sp_externalities::{decl_extension, ExternalitiesExt};
+use sp_runtime_interface::runtime_interface;
+
+/// Contracts pallet outsources debug callbacks through this runtime interface.
+///
+/// Essentially, in our case, it just exposes extensions to the runtime.
+///
+/// At this level, data passed back/forth must be either primitive or implement some specific
+/// traits. For simplicity, we just go with primitives and codec encoded data.
+#[runtime_interface]
+pub trait ContractCallDebugger {
+    fn after_call(
+        &mut self,
+        contract_address: Vec<u8>,
+        is_call: bool,
+        input_data: Vec<u8>,
+        result: Vec<u8>,
+    ) {
+        self.extension::<TracingExt>()
+            .expect("Failed to find `DebugExt` extension")
+            .after_call(contract_address, is_call, input_data, result);
+    }
+
+    fn intercept_call(
+        &mut self,
+        contract_address: Vec<u8>,
+        is_call: bool,
+        input_data: Vec<u8>,
+    ) -> Vec<u8> {
+        self.extension::<InterceptingExt>()
+            .expect("Failed to find `InterceptingExt` extension")
+            .intercept_call(contract_address, is_call, input_data)
+    }
+}
+
+/// This trait describes a runtime extension that can be used to debug contract calls.
+pub trait TracingExtT {
+    /// Called after a contract call is made.
+    fn after_call(
+        &self,
+        _contract_address: Vec<u8>,
+        _is_call: bool,
+        _input_data: Vec<u8>,
+        _result: Vec<u8>,
+    ) {
+    }
+}
+
+decl_extension! {
+    /// A wrapper type for the `TracingExtT` debug extension.
+    pub struct TracingExt(Box<dyn TracingExtT + Send>);
+}
+
+/// This trait describes a runtime extension that can be used to intercept contract calls.
+pub trait InterceptingExtT {
+    /// Called when a contract call is made.
+    ///
+    /// The returned value must be a valid codec encoding for `Option<ExecResult>`.
+    fn intercept_call(
+        &self,
+        _contract_address: Vec<u8>,
+        _is_call: bool,
+        _input_data: Vec<u8>,
+    ) -> Vec<u8> {
+        // By default, do not intercept, continue with the standard procedure.
+        None::<()>.encode()
+    }
+}
+
+decl_extension! {
+    /// A wrapper type for the `InterceptingExtT` debug extension.
+    pub struct InterceptingExt(Box<dyn InterceptingExtT + Send>);
+}
```

Comment:
I wonder why those traits, extensions etc (code above) can't be part of the Substrate. The `CallInterceptor` is already there and abstractions above seem generally applicable.

Reply:
they could be, but I'm worried that pushing these things to Parity will be very painful... I can ask them on Element about it